modules/RTC/CodecSelection.js

```diff
@@ -0,0 +1,207 @@
+        // The call is upgraded to the preferred codec once that list is empty.
+        this.nonPreferredParticipants = [];
+
+        this.conference.on(
+            JitsiConferenceEvents.USER_JOINED,
+            this._onParticipantJoined.bind(this));
+        this.conference.on(
+            JitsiConferenceEvents.USER_LEFT,
+            this._onParticipantLeft.bind(this));
+        this.conference.on(
+            JitsiConferenceEvents._MEDIA_SESSION_STARTED,
+            session => this._onMediaSessionStared(session));
+    }
+
+    /**
+     * Checks if a given string is a valid video codec mime type.
+     *
+     * @param {string} codec the codec string that needs to be validated.
+     * @returns {CodecMimeType|null} mime type if valid, null otherwise.
+     * @private
+     */
+    _getCodecMimeType(codec) {
+        if (typeof codec === 'string') {
+            return Object.values(CodecMimeType).find(value => value === codec.toLowerCase());
+        }
+
+        return null;
+    }
+
+    /**
+     * Checks if the given codec is supported by the browser.
+     *
+     * @param {CodecMimeType} preferredCodec codec to be checked.
+     * @returns {boolean} true if the given codec is supported, false otherwise.
+     * @private
+     */
+    _isCodecSupported(preferredCodec) {
+        // Skip the check on FF and RN because they do not support the getCapabilities API.
+        // It is safe to assume both of them support all the codecs supported by Chrome.
+        if (browser.isFirefox() || browser.isReactNative()) {
+            return true;
+        }
+
+        return window.RTCRtpReceiver
+            && window.RTCRtpReceiver.getCapabilities('video').codecs
+            .some(codec => codec.mimeType.toLowerCase() === `video/${preferredCodec}`);
+    }
+
+    /**
+     * Handles the {@link JitsiConferenceEvents._MEDIA_SESSION_STARTED} event. Codecs need to be
+     * configured on the media session that is newly created.
+     *
+     * @param {JingleSessionPC} mediaSession media session that started.
+     * @returns {void}
+     * @private
+     */
+    _onMediaSessionStared(mediaSession) {
+        const preferredCodec = mediaSession.isP2P ? this.p2pPreferredCodec : this.jvbPreferredCodec;
+        const disabledCodec = this.disabledCodec && this._isCodecSupported(this.disabledCodec)
+            ? this.disabledCodec
+            : null;
+        let codec = preferredCodec;
+
+        // For a new endpoint joining the call, JitsiConferenceEvents.USER_JOINED event is received before the
+        // media session is created, the supported codecs for all the remote endpoints in the call need to be
+        // compared here before setting the codec on the peerconnection.
+        if (!mediaSession.isP2P) {
+            const remoteParticipants = this.conference.getParticipants().map(participant => participant.getId());
+
+            for (const remote of remoteParticipants) {
```

Comment:
instead of maintaining a list of non preferred participants I would probably create a "selector" which selects the codec by going over all remote participants every time any of the input changes. I think it's less error prone to check the state rather than trying to track how things are changing, but it's not a big deal

Reply:
refactored the codec selection logic as per your suggestion.